### PR TITLE
Add "Allow comments by default" in user preferences

### PR DESCRIPTION
**Pull request by alexandre-eliot**
Screenshot 2023-07-27 at 12 57 12
Screenshot 2023-07-27 at 12 57 07

**Untested**

I found myself allowing comments all the time on the solutions I publish.
For convenience, I wish the comments would be enabled by default on every solution I publish.

This branch has for objective to add a comments' preference in the existing user preferences.
This preference would allow comments by default when publishing a solution.

I am sorry I haven't tested it yet. I haven't started the process of setting up the project correctly.
Although my understanding in Ruby is pretty limited, I tried to lay the first stones for implementing this feature, on top of the current user preferences system.

If my code has some flaws or is not working, I hope this PR convinced someone more talented than me, to try and work on this feature.

### CODE REVIEW

File: app/commands/solution/publish.rb [hunk 1]
```diff
@@ -10,7 +10,10 @@ def call
       return if solution.published?
 
       ActiveRecord::Base.transaction do
-        solution.update(published_at: Time.current)
+        solution.update(
+          published_at: Time.current,
+          allow_comments: user_preferences.allow_comments_by_default
```
Comment (iHiD) [inline app/commands/solution/publish.rb]:
```suggestion
          allow_comments: user.preferences.allow_comments_by_default
```

File: app/commands/solution/publish.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 class Solution::Publish
   include Mandate
 
-  initialize_with :solution, :user_track, :iteration_idx
+  initialize_with :solution, :user_track, :user_preferences, :iteration_idx
```
Comment (iHiD) [inline app/commands/solution/publish.rb]:
```suggestion
  initialize_with :solution, :user_track, :iteration_idx
```

File: app/commands/solution/publish.rb (hunk 1 above)
Comment (iHiD) [reply app/commands/solution/publish.rb]: @dem4ron Change `delegate :exercise, to: :solution` to `delegate :exercise, :user, to: :solution` below, and change any `solution.user` to `user` etc

File: config/locales/user_preferences/en.yml [hunk 3]
```diff
@@ -1,3 +1,4 @@
 en:
   user_preferences:
     auto_update_exercises: "Automatically update my solutions to the latest version of exercises if they pass the latest version's tests"
+    allow_comments_by_default: "Allow comments for my published solutions by default"
```
Comment (iHiD) [inline config/locales/user_preferences/en.yml]:
```suggestion
    allow_comments_by_default: "Allow people to comment on my published solutions by default"
```

**Comment by iHiD** [comment]
@alexandre-eliot Thanks for this. @dem4ron is going to get it over the line :)

**Comment by alexandre-eliot** [comment]
@iHiD Thank you for your review as well as for your concern !
@dem4ron Thank you in advance ! :)

**Comment by alexandre-eliot** [comment]
Wow @dem4ron, that looks great ! I'm sorry I failed to see how much work needed to be done to get it over the line. Thanks a lot for your work :)

**Comment by dem4ron** [comment]
> Wow @dem4ron, that looks great ! I'm sorry I failed to see how much work needed to be done to get it over the line. Thanks a lot your work :)

Thanks for your contribution :)

**Comment by iHiD** [comment]
OK. I've added new indexes and were good to go. Thanks both!

**Pull request merged**